### DISCUSSION TITLE
listen on relationship not work

### DISCUSSION THREAD

**Discussion by xdewx** [sqlalchemy/sqlalchemy#8318] [Usage Questions]
### Describe the bug

I have a model like below：
```python
class MyModel(BaseModel):
    students=relationship(Student)
```

I want to format argument students before commit, then I use listen to achieve that. But it seems not work! Could you help me?

```python
def format_students(target, value, oldvalue, initiator):
    if not value:
        return []
    return list(map(lambda item: item if isinstance(item,Student) else Student(**item),value))

listen(MyModel.students,format_students,retval=True)
```

### To Reproduce

```python

```

### Error

format_students never called

```
 AttributeError: 'dict' object has no attribute '_sa_instance_state'
``` 

### Versions

- OS: linux
- Python: 3.9
- SQLAlchemy: 1.4
- Database: sqlite3
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):

**Comment by CaselIT** [comment]
Hi,

There are several thing missing here, like a complete example and a complete strack trace.
Please look at https://github.com/sqlalchemy/sqlalchemy/issues?q=is%3Aissue+label%3A%22great+mcve%22+ for example on how to report a bug

**Reply by xdewx** [reply]
reproduce by following below：
1. git clone https://github.com/whichdew/sqlalchemy-starter.git
2. cd sqlalchemy-starter
3. pytest test_todo.py -s

**Reply by xdewx** [reply]
> reproduce by following below：
> 1. git clone https://github.com/whichdew/sqlalchemy-starter.git
> 2. cd sqlalchemy-starter
> 3. pytest test_todo.py -s

@CaselIT please help me

**Reply by CaselIT** [reply]
Ok not an expert bug set does not seem to be used for relationships. The documentation indicates
> Receive a scalar set event. https://docs.sqlalchemy.org/en/14/orm/events.html#sqlalchemy.orm.AttributeEvents.set

I think you should use instead [`append`](https://docs.sqlalchemy.org/en/14/orm/events.html#sqlalchemy.orm.AttributeEvents.append)

```py
def format_uris2(target, value, initiator):
    print('Todo.uris append', value)
    return value if isinstance(value, Uri) else Uri(**value)
listen(Todo.uris, "append", format_uris2, retval=True)
```

@zzzeek maybe it may make sense better indicate that the `set` event is not fired for relationships?

**Reply by xdewx** [reply]
it works, thank you very much.
